rllib/env/multi_agent_episode.py

```diff
@@ -0,0 +1,535 @@
+        id_: Optional[str] = None,
+        agent_episode_ids: Optional[Dict[str]] = None,
+        *,
+        observations=None,
+        actions=None,
+        rewards=None,
+        states=None,
+        t_started: int = 0,
+        is_terminated: bool = False,
+        is_truncated: bool = False,
+        render_images=None,
+        agent_ids=None,
+        # TODO (simon): Also allow to receive `extra_model_outputs`.
+    ):
+        self.id_ = id_ or uuid.uuid4().hex
+
+        # Agent ids must be provided if data is provided. The Episode cannot
+        # know how many agents are in the environment. Also the number of agents
+        # can grwo or shrink.
+        self.agent_ids: list = [] if agent_ids is None else agent_ids
+
+        # The global last timestep of the episode and the timesteps when this chunk
+        # started.
+        self.t = self.t_started = (
+            t_started if t_started is not None else max(len(observations) - 1, 0)
+        )
+        # Keeps track of the correspondence between agent steps and environment steps.
+        # This is a mapping from agents to `IndexMapping`. The latter keeps
+        # track of the global timesteps at which an agent stepped.
+        self.global_t_to_local_t: Dict[str, List[int]] = self._generate_ts_mapping(
+            observations
+        )
+
+        # Note that all attributes will be recorded along the global timestep
+        # in an multi-agent environment. `SingleAgentEpisodes`
+        self.agent_episodes = {
+            agent_id: self._generate_single_agent_episode(
+                agent_id,
+                observations,
+                actions,
+                rewards,
+                states,
+                agent_episode_ids,
+            )
+            for agent_id in self.agent_ids
+        }
+
+        # obs[-1] is the final observation in the episode.
+        self.is_terminated = is_terminated
+        # obs[-1] is the last obs in a truncated-by-the-env episode (there will no more
+        # observations in following chunks for this episode).
+        self.is_truncated = is_truncated
+        # RGB uint8 images from rendering the env; the images include the corresponding
+        # rewards.
+        assert render_images is None or observations is not None
+        self.render_images = [] if render_images is None else render_images
+
+    def concat_episode(self, episode_chunk: "MultiAgentEpisode") -> None:
+        """Adds the given `episode_chunk` to the right side of self.
+
+        This is used for concatenating ongoing episodes in the buffer with
+        ongoing episodes in the sampler.
+        """
+        assert episode_chunk.id_ == self.id_
+        assert not self.is_done
+        # Make sure the timesteps match.
+        assert self.t == episode_chunk.t_started
+
+        # TODO (simon): Write `validate()` method.
+
+        # Make sure, end matches `episode_chunk`'s beginning for all agents.
+        observations = self.get_observations()
```

Comment:
I like this API thus far. :)

Reply:
Yeah me too. ALso using SAE in MAE - a lot is simply passing through